Added CTRL+Shift+Click Functionality

ldbenoit opened the pull request:
Ctrl+Shift+Click now removes enough currency, one stack at a time, until
you have the required amount for the current tab.
Fixed some currency names.
Fixed incorrect file paths for currencyTradeNames.json and
Currency_All.txt

lemasato commented on POE Trades Companion.ahk:
```diff
@@ -7661,8 +7698,77 @@ GUI_Trades_Update_Tab_Style(tabId) {
+	if (item.name && RegExMatch(clip, "i)(?:" item.name ")[\s\S]*(?:Stack Size: )(\d+)\/(\d+)", match)) {
+		available := match1
+		stackSize := match2
+		; if available amount hasn't changed, it's likely the previous click hasn't gone through yet
```
Causes an issue when the hotkey bugs out and the user puts the currency back where it was.
Example:
- Presses the hotkey. It bugs out by sending shift-ctrl-click which results in showing the in-game "currency splitter".
- User cancels or accepts and put the currency back in its slot, therefore `available` does not change.
- User presses the hotkey again, but nothing happens since `available = lastAvailable`

lemasato commented on POE Trades Companion.ahk:
```diff
@@ -7661,8 +7698,77 @@ GUI_Trades_Update_Tab_Style(tabId) {
+		SendInput {Shift Up}{Ctrl Down}{LButton Down}{LButton Up}{Ctrl Up}
+		return
+	ShiftClickPlus:
+		SendInput {Ctrl Up}{Shift Down}{LButton Down}{LButton Up}{Shift Up}
```
Should probably add a small sleep (200ms?), since the "currency splitter" does not always appear instantly.

lemasato commented on POE Trades Companion.ahk:
```diff
@@ -7661,8 +7698,77 @@ GUI_Trades_Update_Tab_Style(tabId) {
+		available := match1
+		stackSize := match2
+		; if available amount hasn't changed, it's likely the previous click hasn't gone through yet
+		if (available = lastAvailable) {
```
This causes an issue when things bug out and show the currency splitter when it shouldn't, and the user puts the currency back where it was. Example:
- User presses hotkey. It bugs out and show the currency splitter.
- User cancels or take the currency then put in back in place.
- User presses the hotkey again, hoping it works this time but nothing happens since `available = lastAvailable`

lemasato commented on POE Trades Companion.ahk:
```diff
@@ -7661,8 +7698,77 @@ GUI_Trades_Update_Tab_Style(tabId) {
+
+	; Using these because ^{LButton} was finicky, sometimes including shifts or not executing properly
+	CtrlClick:
+		SendInput {Shift Up}{Ctrl Down}{LButton Down}{LButton Up}{Ctrl Up}
```
You can simply use `{Click}` instead of `{LButton Down-Up}`

Sometimes, `{Shift Up}` does not work correctly and show the currency splitter.
Not sure if there is any fix for this, since it's part of the hotkey.

lemasato commented on POE Trades Companion.ahk:
```diff
@@ -7661,8 +7698,77 @@ GUI_Trades_Update_Tab_Style(tabId) {
+		return
+	ShiftClickPlus:
+		SendInput {Ctrl Up}{Shift Down}{LButton Down}{LButton Up}{Shift Up}
+		SendInput, %amount%{Enter}
```
Should add a sleep (200-500ms?) before sending the amount, as the currency splitter panel does not always appear instantly.

lemasato commented on POE Trades Companion.ahk:
```diff
@@ -7661,8 +7698,77 @@ GUI_Trades_Update_Tab_Style(tabId) {
+		stackSize := match2
+		; if available amount hasn't changed, it's likely the previous click hasn't gone through yet
+		if (available = lastAvailable) {
+			return
```
Causes an issue when the hotkey bugs out by showing the currency splitter when it shouldn't. Example:
- User presses the hotkey. It bugs out and shows the currency splitter.
- User cancels or take the currency and put it backs in place.
- User presses the hotkey again, hoping it works out this time but nothing happens since `available = lastAvailable`

lemasato commented on POE Trades Companion.ahk:
```diff
@@ -7661,8 +7698,77 @@ GUI_Trades_Update_Tab_Style(tabId) {
+		return
+	ShiftClickPlus:
+		SendInput {Ctrl Up}{Shift Down}{LButton Down}{LButton Up}{Shift Up}
+		SendInput, %amount%{Enter}
```
Should a sleep delay here (200-500ms?), before sending the amount, since the currency splitter does not always appear instantly.

ldbenoit replied:
I've never actually had it not show up. Just tested around 100 times, is this something you have experienced?

lemasato replied:
Disregard. It was due to the shift key issue.
I thought it was the ShfitClickPlus label messing up, but it was actually the CtrlClick label not releasing the Shift key.

lemasato commented on POE Trades Companion.ahk:
```diff
@@ -7661,8 +7698,77 @@ GUI_Trades_Update_Tab_Style(tabId) {
\ No newline at end of file
+#Include BetaFuncs.ahk
+
+#IfWinActive ahk_class POEWindowClass
```
ahk_group POEGame

ldbenoit commented:
I changed ahk_class to ahk_group and changed the key sequence a bit. I hadn't actually had the shift key  mess up in a bit so I assumed it was fixed. I changed it so {Shift Up} happens right before the mouse key, can you see if that helps? I couldn't get it to happen after I changed that. 
It did just occur to me that if you are using it on single stacks the lastAvailable would bug out as well. I'm not sure the best fix for that.

ldbenoit commented on POE Trades Companion.ahk:
```diff
@@ -7716,9 +7716,9 @@ StackClick() {
-		stackSize := match2
+	if (item.name && RegExMatch(clip, "i)(?:" item.name ")[\s\S]*(?:Stack Size: )(\d+(?:,\d+)*)\/(\d+(?:,\d+)*)", match)) {
+		available := RegexReplace(match1, ",")
+		stackSize := RegexReplace(match2, ",")
```
Hmm, if I'm understanding correctly poe is localized to show either commas or periods as separators in large numbers. Would this work for you?

`if (item.name && RegExMatch(clip, "i)(?:" item.name ")[\s\S]*(?:Stack Size: )(\d+(?:[,.]\d+)*)\/(\d+(?:[,.]\d+)*)", match)) {`
`available := RegexReplace(match1, "[,.]")`
`stackSize := RegexReplace(match2, "[,.]")`

ldbenoit commented:
> After fiddling around, I found the perfect fix: Hotkey, *Shift, DoNothing, On
> It will prevent the Shift modifier from interfering with the inputs. It may be a good idea to do the same for Ctrl, and add them in both ClickClick: and ShiftClickPlus:, just "in case".

It's strange we are having such different experiences. I just tried that method and it was causing the shift+click to happen quite frequently, maybe one in 3-5 clicks.
Using SendInput {Ctrl Down}{Shift Up}{LButton Up}{Ctrl Up} it happens once every few dozen, maybe 100 clicks. Maybe I did something wrong but it wasn't working for me.

I made some changes to the regex to allow for either '.' or ',' as separators. I also noticed 'Stack Size:' might be localized as well and really wasn't doing anything so I removed it.

lemasato commented:
What's even more than weird is that the macro works just fine when using the fix above.
But when using this standalone script ([CtrlClick.zip](https://github.com/lemasato/POE-Trades-Companion/files/1275100/CtrlClick.zip)), it simply doesn't work... Still haven't found why.

ldbenoit commented:
I tried that out and it works well. I think that tiny delay makes it work for me. I have added it in my latest commit, let me know if it works for you.
I also reset lastAvailable when transfering individual stacks to not cause problems. It now has a 250ms delay to reduce the chance of you clicking twice on the same stack and messing up the count. Can you think of a better fix for this?

lemasato commented:
Strange... The shift fix does work. But under specific circonstancies, it will bug out:
- `az @From iSellStuff: Hi, I'd like to buy your 43 alteration for my 1 chaos in Standard`
- Press Ctrl+Shift+Click three times. The two first times do ctrl-click and the third will do shift-click, as expected. Close the tab.
- `az @From iSellStuff: Hi, I'd like to buy your 43 alteration for my 1 chaos in Standard`
- Ctrl+Shift+Click will work on the first time. Will bug on the second. Works fine on third time.
- Close the tab.
- `az @From iSellStuff: Hi, I'd like to buy your 43 alteration for my 1 chaos in Standard`
- From now on, the first click will bug out, but second and third will work just fine.

It's really really obscure. When disabling `GUI_Trades_Inc_WithdrawTally(tabId, amount) at Line 7749`, you can Ctrl+Shift+Click as many times as you want without CtrlClick: bugging out.
I tried disabling ShiftClickPlus:, but it still happens. It's not related.

Adding **{LButton Up}** before **{Click}** and running **Hotkey, \*Shift/\*Ctrl/LButton, DoNothing, Off**
before **Send {Shift %shiftState%}{Ctrl %ctrlState%}**, the issue still happens. BUT the currency still DOES get moved to your inventory.

***

Also the `available = lastAvailable` issue remains:
- `az @From iSellStuff: Hi, I'd like to buy your 43 alteration for my 1 chaos in Standard`
- Press Ctrl+Shift+Click three times. The two first times do ctrl-click and the third will do shift-click, as expected.
- Press Ctrl+Shift+Click one additional time. Close the tab. 
- `az @From iSellStuff: Hi, I'd like to buy your 43 alteration for my 1 chaos in Standard`
- Ctrl+Shift+Click will not do anything.

My fix for this:
In **`if ((required - withdrawn) <= 0) {`** add **`lastAvailable := 0`**

***

> I also reset lastAvailable when transfering individual stacks so it doesn't cause problems for the next click. It now has a 250ms delay to reduce the chance of you clicking twice on the same stack and messing up the count. Can you think of a better fix for this?

Nice fix, I didn't think about individual stacks.

ldbenoit commented:
That is really strange and can confirm it happens for me as well. I found just replacing {Click} with {Lbutton Up} fixes it for me.

lemasato commented on POE Trades Companion.ahk:
```diff
@@ -7661,8 +7698,130 @@ GUI_Trades_Update_Tab_Style(tabId) {
+	SendInput {Shift Up}^c
+
+	; wait for clipboard to be populated and do nothing if it fails
+	ClipWait, .2
```
I see you are using **`ClipWait, .2`** here but **`.1`** in the StackClick function.
I personally haven't encountered any issue using **`.1`**, so it's most likely safe to use.
Though, I think it may cause issues with external clipboard managers (See #79). Yet have to test.

ldbenoit replied:
I think my thought was that since you wouldn't be spamming this hotkey I would give it a little more lenience. It can probably be changed to .1 as I don't think it causes any problems.

lemasato commented on POE Trades Companion.ahk:
```diff
@@ -7661,8 +7698,130 @@ GUI_Trades_Update_Tab_Style(tabId) {
+		Send {Shift %shiftState%}{Ctrl %ctrlState%} ; Restore ctrl/shift state
+		return
+	Finished: 
+		SoundPlay,% ProgramSettings.Whisper_Sound_Path
```
Now that the function is reaching its final stage, maybe we could switch to using a ToolTip here?
And remove it upon moving the mouse by using **`Remove_Tooltip_OnMouseMove()`**

ldbenoit replied:
I've added tooltips to the latest commit, although I like having a sound as well.

lemasato commented on POE Trades Companion.ahk:
```diff
@@ -7661,8 +7698,130 @@ GUI_Trades_Update_Tab_Style(tabId) {
+		stackSize := RegexReplace(match2, "[,.]")
+		; if available amount hasn't changed, it's likely the previous click hasn't gone through yet
+		if (available = lastAvailable) {
+			return
```
Maybe use a ToolTip aswell? So the user can have an idea of what's going on and why the macro was cancelled.

lemasato commented on POE Trades Companion.ahk:
```diff
@@ -7661,8 +7698,130 @@ GUI_Trades_Update_Tab_Style(tabId) {
+
+		; Don't do anything if we've already withdrawn all we need
+		if ((required - withdrawn) <= 0) {
+			return
```
Yet another ToolTip?

lemasato commented:
Thank you for your work!
